### PR TITLE
Fix core logic

### PR DESCRIPTION
**Pull request by hallazzang**

### CODE REVIEW

File: x/farming/types/keys.go [hunk 1]
```diff
@@ -122,7 +113,20 @@ func GetRewardByFarmerAddrIndexPrefix(farmerAcc sdk.AccAddress) []byte {
...
+func ParseRewardKey(key []byte) (stakingCoinDenom string, farmer sdk.AccAddress, err error) {
+	denomLen := key[1]
+	stakingCoinDenom = string(key[2 : 2+denomLen])
+	farmer = key[2+denomLen+1:]
+	// TODO: add error case
+	return
+}
+
+// ParseRewardByFarmerAddrIndexKey parses a key of RewardByFarmerAddrIndex from bytes.
+func ParseRewardByFarmerAddrIndexKey(key []byte) (stakingCoinDenom string, farmer sdk.AccAddress, err error) {
```
Comment (dongsam) [inline x/farming/types/keys.go:126]: How about the arguments ordering to `func ParseRewardByFarmerAddrIndexKey(key []byte) (farmer sdk.AccAddress, stakingCoinDenom string, err error)` to consistent with indexing ordering

File: x/farming/types/keys.go [hunk 2]
```diff
@@ -27,9 +27,9 @@ const (
 
 var (
 	// param key for global farming plan IDs
-	GlobalPlanIdKey           = []byte("globalPlanId")
+	GlobalPlanIDKey           = []byte("globalPlanId")
 	GlobalLastEpochTimePrefix = []byte("globalLastEpochTime")
-	GlobalStakingIdKey        = []byte("globalStakingId")
+	GlobalLastStakingIDKey    = []byte("globalLastStakingId")
```
Comment (dongsam) [inline x/farming/types/keys.go:32]: `GlobalPlanIdKey ` and `GlobalStakingIdKey ` operate in the same form, but consistency is broken to attach `Last` only to `GlobalStakingIdKey `. It also needs an update on Spec to change.

File: x/farming/keeper/keeper.go [hunk 3]
```diff
@@ -76,7 +76,7 @@ func (k Keeper) GetNextPlanIDWithUpdate(ctx sdk.Context) uint64 {
 	var id uint64
 	store := ctx.KVStore(k.storeKey)
 
-	bz := store.Get(types.GlobalPlanIdKey)
+	bz := store.Get(types.GlobalPlanIDKey)
```
Comment (dongsam) [inline x/farming/keeper/keeper.go:79]: More discussion is needed, but why don't we unify it with `Id` to match the field name generated by proto?

File: proto/tendermint/farming/v1beta1/farming.proto [hunk 4]
```diff
@@ -157,15 +157,9 @@ message Reward {
   option (gogoproto.goproto_getters)  = false;
   option (gogoproto.goproto_stringer) = false;
 
-  // farmer defines the bech32-encoded address of the staker for the plan
-  string farmer = 1;
-
-  // staking_coin_denom is denom of staked coin as a source of the reward
-  string staking_coin_denom = 2 [(gogoproto.moretags) = "yaml:\"staking_coin_denom\""];
-
```
Comment (dongsam) [inline proto/tendermint/farming/v1beta1/farming.proto:165]: We need to original `Reward` struct for export/import genesis  
Now setting original `Reward` itself on kv, but I'll do research later and add it when saving `RewardCoin` only.

**Review by dongsam — CHANGES_REQUESTED** [review]

File: x/farming/keeper/staking.go [hunk 5]
```diff
@@ -44,51 +80,24 @@ func (k Keeper) GetAllStakings(ctx sdk.Context) (stakings []types.Staking) {
...
 func (k Keeper) SetStaking(ctx sdk.Context, staking types.Staking) {
 	store := ctx.KVStore(k.storeKey)
 	bz := k.cdc.MustMarshal(&staking)
 	store.Set(types.GetStakingKey(staking.Id), bz)
 }
 
-// RemoveStaking removes an staking for the staking mapper store.
-func (k Keeper) RemoveStaking(ctx sdk.Context, staking types.Staking) {
+// DeleteStaking deletes a staking.
+func (k Keeper) DeleteStaking(ctx sdk.Context, id uint64) {
```
Comment (dongsam) [inline x/farming/keeper/staking.go:91]: Need to get staking object for remove staking index by farmer and denom

File: x/farming/types/keys.go (hunk 1 above)
Comment (hallazzang) [reply x/farming/types/keys.go:126]: Ok, it makes sense.

File: x/farming/types/keys.go (hunk 2 above)
Comment (hallazzang) [reply x/farming/types/keys.go:32]: I'll revert back to the original name

File: x/farming/keeper/keeper.go (hunk 3 above)
Comment (hallazzang) [reply x/farming/keeper/keeper.go:79]: Ok.

File: x/farming/keeper/reward.go [hunk 6]
```diff
@@ -132,23 +102,42 @@ func (k Keeper) UnmarshalReward(bz []byte) (types.Reward, error) {
...
 	}
 
 	for _, denom := range stakingCoinDenoms {
-		k.SetReward(ctx, denom, farmer, sdk.NewCoins())
+		k.DeleteReward(ctx, denom, farmerAcc)
 	}
 
-	return amount, nil
+	if len(k.GetRewardsByFarmer(ctx, farmerAcc)) == 0 {
+		k.GetStakingIDByFarmer(ctx, farmerAcc)
```
Comment (dongsam) [inline x/farming/keeper/reward.go:124]: `k.GetStakingIDByFarmer(ctx, farmerAcc)` is unused with result
and `GetStakingIDByFarmer` can be refactoring to `GetStakingIdByFarmer`

File: x/farming/keeper/reward.go [hunk 7]
```diff
@@ -18,29 +20,9 @@ func (k Keeper) GetReward(ctx sdk.Context, stakingCoinDenom string, farmerAcc sd
...
-
-// GetRewardsByStakingCoinDenom reads from kvstore and return a specific Reward indexed by given staking coin denom
-func (k Keeper) GetRewardsByStakingCoinDenom(ctx sdk.Context, denom string) (rewards []types.Reward) {
-	k.IterateRewardsByStakingCoinDenom(ctx, denom, func(reward types.Reward) bool {
-		rewards = append(rewards, reward)
-		return false
-	})
-
-	return rewards
-}
```
Comment (dongsam) [inline x/farming/keeper/reward.go:39]: It think `GetAllRewards` and `GetRewardsByStakingCoinDenom` will be used other core logics

File: x/farming/keeper/reward.go [hunk 8]
```diff
@@ -51,24 +33,16 @@ func (k Keeper) GetRewardsByFarmer(ctx sdk.Context, farmer sdk.AccAddress) (rewa
...
 	store := ctx.KVStore(k.storeKey)
-	// TODO: only rewardCoins
-	// 0x31 | StakingCoinDenomAddrLen (1 byte) | StakingCoinDenom | FarmerAddrLen (1 byte) | FarmerAddr -> ProtocolBuffer(sdk.Coins) RewardCoins
-	reward := types.Reward{
-		Farmer:           farmerAcc.String(),
-		StakingCoinDenom: stakingCoinDenom,
-		RewardCoins:      rewardCoins,
-	}
-	bz := k.cdc.MustMarshal(&reward)
+	bz := k.cdc.MustMarshal(&types.Reward{RewardCoins: rewardCoins})
```
Comment (dongsam) [inline x/farming/keeper/reward.go:36]: I think we need to marshal as `types.RewardCoins` type to unify with unmarshal logic

**Review by dongsam — CHANGES_REQUESTED** [review]
How about unifying to `Id` from `ID` from every struct and function names

**Pull request merged**